src/main/java/net/kaczmarzyk/spring/data/jpa/web/EnhancerUtil.java

```diff
@@ -15,33 +15,74 @@
  */
 package net.kaczmarzyk.spring.data.jpa.web;
 
-import java.lang.reflect.Method;
-
 import org.springframework.cglib.proxy.Enhancer;
 import org.springframework.cglib.proxy.MethodInterceptor;
 import org.springframework.cglib.proxy.MethodProxy;
 import org.springframework.data.jpa.domain.Specification;
 
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+
 
 /**
  * @author Tomasz Kaczmarzyk
  */
 class EnhancerUtil {
 
-    @SuppressWarnings("unchecked")
+	@SuppressWarnings("unchecked")
 	static <T> T wrapWithIfaceImplementation(final Class<T> iface, final Specification<Object> targetSpec) {
-    	Enhancer enhancer = new Enhancer();
-		enhancer.setInterfaces(new Class[] { iface });
+		Enhancer enhancer = new Enhancer();
+		enhancer.setInterfaces(new Class[]{ iface });
 		enhancer.setCallback(new MethodInterceptor() {
-            @Override
-            public Object intercept(Object obj, Method method, Object[] args, MethodProxy proxy) throws Throwable {
-            	if ("toString".equals(method.getName())) {
-            		return iface.getSimpleName() + "[" + proxy.invoke(targetSpec, args) + "]";
-            	}
-            	return proxy.invoke(targetSpec, args);
-            }
-        });
-		
+			@Override
+			public Object intercept(Object obj, Method method, Object[] args, MethodProxy proxy) throws Throwable {
+				if ("toString".equals(method.getName())) {
+					return iface.getSimpleName() + "[" + proxy.invoke(targetSpec, args) + "]";
+				}
+
+				if ("equals".equals(method.getName())) {
+					if (args.length != 1
+							|| args[0] == null
+							|| args[0].getClass().getInterfaces().length == 0
+							|| obj.getClass().getInterfaces().length == 0) {
+						return false;
+					}
+
+					if (!args[0].getClass().getInterfaces()[0].equals(obj.getClass().getInterfaces()[0])) {
+						return false;
+					}
+
+					return ReflectionUtils.get(ReflectionUtils.get(args[0], "CGLIB$CALLBACK_0"), "val$targetSpec").equals(targetSpec);
```

Comment:
I'm not sure if I see it correctly but it looks to me that you are assuming that the arg[0] is a cglib-enhanced object. It does not have to be the case (?). (let's discuss)

Reply:
On the second thought.... I think you assumed that classes have to be the same -- otherwise equals should return false anyway. BUT you are checking interfaces only in the above check. So theoretically it might happen that completely different classes (implementing the same interface) would be compared here.

I think that at least a check "is it a cglib class implementing this interface" would be good.
OR even better -- just a rough idea -- let's add some method/marked to all proxies, so that it is easy to determine that an object is a proxy created by SAR. E.g. you could add an additional marker interface or sth like that.  We need to think about it but genrally we must be really sure what we are comparing with.